node/pkg/watchers/evm/watcher.go

```diff
@@ -982,197 +1010,461 @@ func (w *Watcher) ccqHandleQuery(logger *zap.Logger, ctx context.Context, queryR
+
+	if err != nil {
+		logger.Error("failed to process eth_call_by_timestamp query request",
+			zap.Error(err),
+			zap.String("block", block),
+			zap.Any("batch", batch),
+		)
+		w.ccqSendQueryResponseForError(logger, queryRequest, query.QueryRetryNeeded)
+		return
+	}
+
+	if blockError != nil {
+		logger.Error("failed to process eth_call_by_timestamp query block request",
+			zap.Error(blockError),
+			zap.String("block", block),
+			zap.Any("batch", batch),
+		)
+		w.ccqSendQueryResponseForError(logger, queryRequest, query.QueryRetryNeeded)
+		return
+	}
+
+	if blockResult.Number == nil {
+		logger.Error("invalid eth_call_by_timestamp query block result",
+			zap.String("eth_network", w.networkName),
+			zap.String("block", block),
+			zap.Any("batch", batch),
+		)
+		w.ccqSendQueryResponseForError(logger, queryRequest, query.QueryRetryNeeded)
+		return
+	}
+
+	if blockResult.Number.ToInt().Cmp(w.ccqMaxBlockNumber) > 0 {
+		logger.Error("block number too large for eth_call_by_timestamp",
+			zap.String("eth_network", w.networkName),
+			zap.String("block", block),
+			zap.Any("batch", batch),
+		)
+		w.ccqSendQueryResponseForError(logger, queryRequest, query.QueryRetryNeeded)
+		return
+	}
+
+	resp := query.EthCallByTimestampQueryResponse{
+		TargetBlockNumber: blockResult.Number.ToInt().Uint64(),
+		TargetBlockHash:   blockResult.Hash,
+		TargetBlockTime:   time.Unix(int64(blockResult.Time), 0),
+		// TODO: Fill in the followings!!!
+		Results: [][]byte{},
+	}
+
+	errFound := false
+	for idx := range req.CallData {
+		if evmCallData[idx].callErr != nil {
+			logger.Error("failed to process eth_call_by_timestamp query call request",
+				zap.Error(evmCallData[idx].callErr),
 				zap.String("block", block),
+				zap.Int("errorIdx", idx),
 				zap.Any("batch", batch),
 			)
-			w.ccqSendQueryResponse(logger, queryRequest, query.QueryRetryNeeded, nil)
-			return
+			w.ccqSendQueryResponseForError(logger, queryRequest, query.QueryRetryNeeded)
+			errFound = true
+			break
 		}
 
-		if blockResult.Number.ToInt().Cmp(ccqMaxBlockNumber) > 0 {
-			logger.Error("block number too large",
+		// Nil or Empty results are not valid
+		// eth_call will return empty when the state doesn't exist for a block
+		if len(*evmCallData[idx].callResult) == 0 {
+			logger.Error("invalid call result for eth_call_by_timestamp",
 				zap.String("eth_network", w.networkName),
 				zap.String("block", block),
+				zap.Int("errorIdx", idx),
 				zap.Any("batch", batch),
 			)
-			w.ccqSendQueryResponse(logger, queryRequest, query.QueryRetryNeeded, nil)
-			return
+			w.ccqSendQueryResponseForError(logger, queryRequest, query.QueryRetryNeeded)
+			errFound = true
+			break
 		}
 
-		resp := query.EthCallQueryResponse{
-			BlockNumber: blockResult.Number.ToInt().Uint64(),
-			Hash:        blockResult.Hash,
-			Time:        time.Unix(int64(blockResult.Time), 0),
-			Results:     [][]byte{},
-		}
+		/*
+			target_block.timestamp <= target_time < following_block.timestamp
```

Comment:
I think this check only needs to be performed once, perhaps above the response building / after the block error checking